qp_shotgun/kneaddata/kneaddata.py

```diff
@@ -0,0 +1,287 @@
+            continue
+        elif value is False:
+            continue
+        elif value:
+            params.append('--%s %s' % (param, value))
+            continue
+
+    param_string = ' '.join(params)
+
+    return(param_string)
+
+
+def generate_kneaddata_commands(forward_seqs, reverse_seqs, map_file,
+                                out_dir, parameters):
+    """Generates the KneadData commands
+
+    Parameters
+    ----------
+    forward_seqs : list of str
+        The list of forward seqs filepaths
+    reverse_seqs : list of str
+        The list of reverse seqs filepaths
+    map_file : str
+        The path to the mapping file
+    out_dir : str
+        The job output directory
+    parameters : dict
+        The command's parameters, keyed by parameter name
+
+    Returns
+    -------
+    cmds: list of str
+        The KneadData commands
+    samples: list of tup
+        list of 3-tuples with run prefix, sample name, fwd read fp, rev read fp
+
+    Raises
+    ------
+    ValueError
+        If the rev is not an empty list and the same length than fwd seqs
+        The prefixes of the run_prefix don't match the file names
+
+    Notes
+    -----
+    Currently this is requiring matched pairs in the make_read_pairs_per_sample
+    step but implicitly allowing empty reverse reads in the actual command
+    generation. This behavior may allow support of situations with empty
+    reverse reads in some samples, for example after trimming and QC.
+    """
+    # making sure the forward and reverse reads are in the same order
+
+    # we match filenames, samples, and run prefixes
+    samples = make_read_pairs_per_sample(forward_seqs, reverse_seqs, map_file)
+
+    cmds = []
+
+    param_string = format_kneaddata_params(parameters)
+    prefixes = []
+    for run_prefix, sample, f_fp, r_fp in samples:
+        prefixes.append(run_prefix)
+        if r_fp is None:
+            cmds.append('kneaddata --input "%s" --output "%s" '
+                        '--output-prefix "%s" %s' %
+                        (f_fp, join(out_dir, run_prefix),
+                         run_prefix, param_string))
+        else:
+            cmds.append('kneaddata --input "%s" --input "%s" --output "%s" '
+                        '--output-prefix "%s" %s' %
+                        (f_fp, r_fp, join(out_dir, run_prefix),
+                         run_prefix, param_string))
+
+    return cmds, samples
+
+
+def _run_commands(qclient, job_id, commands, msg):
+    for i, cmd in enumerate(commands):
+        qclient.update_job_step(job_id, msg % i)
+        std_out, std_err, return_value = system_call(cmd)
+        if return_value != 0:
+            error_msg = ("Error running KneadData:\nStd out: %s\nStd err: %s"
+                         "\n\nCommand run was:\n%s"
+                         % (std_out, std_err, cmd))
+            return False, error_msg
+
+    return True, ""
+
+
+def _per_sample_ainfo(out_dir, samples):
```

Comment:
This is adding 1 artifact per output file - I don't think this is the desired behavior right? I think what we want to get is a single artifact with all the sequence file. Note that the number of output artifacts from a command should be deterministic (i.e. it is known in advance how many artifacts a command is going to generate). Note that artifact != file so it shouldn't be a problem.

Now, extending on that, it looks like we can potentially group the output files in 3 different artifacts: (1) clean paired reads with R1 and R2, (2) clean unpaired R1, and (3) clean unpaired R2. However, this doesn't work for the case in which R2 is not provided, since we only have 1 output artifact. To solve this case, I would add 2 commands, 1 for paired reads and 1 for unpaired reads. Note that internally you can be using the same functions if you want, but from the Qiita point of view this simplifies things as the output artifacts are known in advance.

To create the 3 artifacts for the paired case, the code would be something like:
```python
paired_files = []
r1_files = []
r2_files = []
for run_prefix, sample, f_fp, r_fp in samples:
    sam_out_dir = join(out_dir, run_prefix)
    paired_files.append((join(sam_out_dir, '%s_paired_1.fastq' % run_prefix), 'preprocessed_fastq'))
    paired_files.append((join(sam_out_dir, '%s_paired_2.fastq' % run_prefix), 'preprocessed_fastq'))
    r1_files.append((join(sam_out_dir, '%s_unmatched_1.fastq' % run_prefix), 'preprocessed_fastq'))
    r2_files.append((join(sam_out_dir, '%s_unmatched_2.fastq' % run_prefix), 'preprocessed_fastq'))

ainfo = [ArtifactInfo('clean paired', 'per_sample_FASTQ', paired_files),
         ArtifactInfo('clean unmatched R1', 'per_sample_FASTQ', r1_files),
         ArtifactInfo('clean unmatched R1', 'per_sample_FASTQ', r2_files)
```

Note that the definition of the command in the `__init__.py` file would need to be update to show that this command outputs 3 artifacts.